compose/config/config.py

```diff
@@ -165,13 +164,11 @@ def find_candidates_in_parent_dirs(filenames, path):
     return (candidates, path)
 
 
-@validate_top_level_object
-@validate_service_names
 def pre_process_config(config):
-    """
-    Pre validation checks and processing of the config file to interpolate env
+    """Pre validation checks and processing of the config file to interpolate env
     vars returning a config dict ready to be tested against the schema.
     """
+    validate_top_level_object(config)
```

Comment:
Given the direction you're going with this approach, I don't think it makes sense to have this being called as part of `pre_process_config`. As you've already called `validate_top_level_object` for each config file, so there's no need to include it in this function and call it again.